arwen/contexts/storage.go

```diff
@@ -112,7 +119,7 @@ func (context *storageContext) SetStorage(key []byte, value []byte) int32 {
 		return int32(arwen.StorageAdded)
 	}
 	if bytes.Equal(value, zero) {
-		freeGas := metering.GasSchedule().BaseOperationCost.StorePerByte * uint64(len(oldValue))
+		freeGas := metering.GasSchedule().BaseOperationCost.ReleasePerByte * uint64(len(oldValue))
 		metering.FreeGas(freeGas)
 		return int32(arwen.StorageDeleted)
 	}
```

Comment:
Right, this needs to be handled. But should we allow a "free write to storage" if the new value is shorter than the old value? 

We definitely must release the gas for the difference, but we could still consume gas for persisting the shorter new value, because it will be written to storage. But overall, doing this would result in a lower total cost for `storageStore()` (i.e. `gasForWritingNewValue` - `gasReleasedByDifference`).

This would cover the case where a SC persisting 1000 values of storage at full cost, each 100 bytes in size, but later just overwriting them with a single different byte. Because the new and old values are the same length, this can be abused with very little gas cost, but higher resource cost for the node.

Reply:
you already took the cost of calling the save Method :)

Reply:
Indeed. Changed to only consume the extra difference, or to release gas if the new value is shorter.